File: saunity/src/main/java/tv/superawesome/plugins/unity/SAUnityVideoAd.java
/**
 * @Copyright:   SuperAwesome Trading Limited 2017
 * @Author:      Gabriel Coman (gabriel.coman@superawesome.tv)
 */
package tv.superawesome.plugins.unity;

import android.content.Context;

import tv.superawesome.lib.sasession.SAConfiguration;
import tv.superawesome.sdk.views.SAEvent;
import tv.superawesome.sdk.views.SAInterface;
import tv.superawesome.sdk.views.SAOrientation;
import tv.superawesome.sdk.views.SAVideoAd;

/**
 * Class that holds a number of static methods used to communicate with Unity
 */
public class SAUnityVideoAd {

    private static final String unityName = "SAVideoAd";

    /**
     * Method that creates a new Video Ad (from Unity)
     */
    public static void SuperAwesomeUnitySAVideoAdCreate (Context context) {
        SAVideoAd.setListener(new SAInterface() {
            @Override
            public void onEvent(int placementId, SAEvent event) {
                switch (event) {
                    case adLoaded: SAUnityCallback.sendAdCallback(unityName, placementId, "adLoaded"); break;
                    case adFailedToLoad: SAUnityCallback.sendAdCallback(unityName, placementId, "adFailedToShow"); break;
                    case adShown: SAUnityCallback.sendAdCallback(unityName, placementId, "adShown"); break;
                    case adFailedToShow: SAUnityCallback.sendAdCallback(unityName, placementId, "adFailedToShow"); break;
                    case adClicked: SAUnityCallback.sendAdCallback(unityName, placementId, "adClicked"); break;
                    case adClosed: SAUnityCallback.sendAdCallback(unityName, placementId, "adClosed"); break;
                }
            }
        });
    }

    /**
     * Method that loads a new Video Ad (from Unity)
     */
    public static void SuperAwesomeUnitySAVideoAdLoad(Context context, int placementId, int configuration, boolean test) {
        SAVideoAd.setTestMode(test);
        SAVideoAd.setConfiguration(SAConfiguration.fromValue(configuration));
        SAVideoAd.load(placementId, context);
    }

    /**
     * Method that checks to see if an ad is available for a video ad (from Unity)
     */
    public static boolean SuperAwesomeUnitySAVideoAdHasAdAvailable (Context context, int placementId) {
        return SAVideoAd.hasAdAvailable(placementId);
    }

    /**
     * Method that plays a new video ad (from Unity)
     */
    public static void SuperAwesomeUnitySAVideoAdPlay (Context context, int placementId, boolean isParentalGateEnabled, boolean shouldShowCloseButton, boolean shouldShowSmallClickButton, boolean shouldAutomaticallyCloseAtEnd, int orientation, boolean isBackButtonEnabled) {
        SAVideoAd.setParentalGate(isParentalGateEnabled);
        SAVideoAd.setCloseAtEnd(shouldAutomaticallyCloseAtEnd);
        SAVideoAd.setCloseButton(shouldShowCloseButton);
        SAVideoAd.setSmallClick(shouldShowSmallClickButton);
        SAVideoAd.setBackButton(isBackButtonEnabled);
        SAVideoAd.setOrientation(SAOrientation.fromValue(orientation));
        SAVideoAd.play(placementId, context);
    }

}

File: superawesome-unity/src/main/java/tv/superawesome/plugins/publisher/unity/SAUnityVideoAd.kt
/**
 * @Copyright:   SADefaults Trading Limited 2017
 * @Author:      Gabriel Coman (gabriel.coman@superawesome.tv)
 */
package tv.superawesome.plugins.publisher.unity

import android.content.Context
import tv.superawesome.sdk.publisher.common.models.AdRequest
import tv.superawesome.sdk.publisher.common.models.Orientation
import tv.superawesome.sdk.publisher.common.models.SAEvent
import tv.superawesome.sdk.publisher.common.models.SAInterface
import tv.superawesome.sdk.publisher.common.ui.video.SAVideoAd

/**
 * Class that holds a number of static methods used to communicate with Unity
 */
object SAUnityVideoAd {
    private const val unityName = "SAVideoAd"

    /**
     * Method that creates a new Video Ad (from Unity)
     */
    fun SuperAwesomeUnitySAVideoAdCreate(context: Context?) {
        SAVideoAd.setListener { placementId, event ->
            when (event) {
                SAEvent.AdLoaded -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdLoaded.toString()
                )
                SAEvent.AdEmpty -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdEmpty.toString()
                )
                SAEvent.AdFailedToLoad -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdFailedToLoad.toString()
                )
                SAEvent.AdAlreadyLoaded -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdAlreadyLoaded.toString()
                )
                SAEvent.AdShown -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdShown.toString()
                )
                SAEvent.AdFailedToShow -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdFailedToShow.toString()
                )
                SAEvent.AdClicked -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdClicked.toString()
                )
                SAEvent.AdEnded -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdEnded.toString()
                )
                SAEvent.AdClosed -> SAUnityCallback.sendAdCallback(
                    unityName,
                    placementId,
                    SAEvent.AdClosed.toString()
                )
            }
        }
    }

    /**
     * Method that loads a new Video Ad (from Unity)
     */
    fun SuperAwesomeUnitySAVideoAdLoad(
            context: Context,
            placementId: Int,
            configuration: Int,
            test: Boolean,
            playback: Int,
    ) {
        SAVideoAd.setTestMode(test)
        AdRequest.StartDelay.fromValue(playback)?.let { SAVideoAd.setPlaybackMode(it) }
        SAVideoAd.load(placementId, context)
    }

    /**
     * Method that checks to see if an ad is available for a video ad (from Unity)
     */
    fun SuperAwesomeUnitySAVideoAdHasAdAvailable(context: Context?, placementId: Int): Boolean =
            SAVideoAd.hasAdAvailable(placementId)

    /**
     * Method that plays a new video ad (from Unity)
     */
    fun SuperAwesomeUnitySAVideoAdPlay(
            context: Context,
            placementId: Int,
            isParentalGateEnabled: Boolean,
            isBumperPageEnabled: Boolean,
            shouldShowCloseButton: Boolean,
            shouldShowSmallClickButton: Boolean,
            shouldAutomaticallyCloseAtEnd: Boolean,
            orientation: Int,
            isBackButtonEnabled: Boolean,
            shouldShowCloseWarning: Boolean,
    ) {
        SAVideoAd.setParentalGate(isParentalGateEnabled)
        SAVideoAd.setBumperPage(isBumperPageEnabled)
        SAVideoAd.setCloseAtEnd(shouldAutomaticallyCloseAtEnd)
        SAVideoAd.setCloseButton(shouldShowCloseButton)
        SAVideoAd.setSmallClick(shouldShowSmallClickButton)
        SAVideoAd.setBackButton(isBackButtonEnabled)
        Orientation.fromValue(orientation)?.let { SAVideoAd.setOrientation(it) }
        // TODO: update close warning after merge with master branch
        //SAVideoAd.setCloseButtonWarning(shouldShowCloseWarning)
        SAVideoAd.play(placementId, context)
    }
}